provider/kubernetes/kubernetes.go

```diff
@@ -677,3 +706,64 @@ func getRateLimit(i *extensionsv1beta1.Ingress) *types.RateLimit {
 
 	return rateLimit
 }
+
+func getServicesPercentageWeights(i *extensionsv1beta1.Ingress) (map[string]*percentageValue, error) {
+	rawPercentageWeights := getStringValue(i.Annotations, annotationKubernetesPercentageWeights, "")
+
+	percentageWeight := make(map[string]string)
+
+	if err := yaml.Unmarshal([]byte(rawPercentageWeights), percentageWeight); err != nil {
+		return nil, err
+	}
+
+	servicesPercentageWeights := make(map[string]*percentageValue)
+	for serviceName, percentageStr := range percentageWeight {
+		percentageValue, err := newPercentageValueFromString(percentageStr)
+		if err != nil {
+			return nil, fmt.Errorf("invalid percentage value %q in ingress %s: %v", percentageStr, i.Name, err)
+		}
+
+		servicesPercentageWeights[serviceName] = percentageValue
+	}
+	return servicesPercentageWeights, nil
+}
+
+func getLeftFraction(k8sClient Client, namespace string, paths []extensionsv1beta1.HTTPIngressPath, servicesPercentageWeights map[string]*percentageValue) (map[string]*percentageValue, map[string]int, error) {
+	leftFractionPercentages := make(map[string]*percentageValue)
+	leftFractionInstanceCounts := make(map[string]int)
+	oneHundredPercentageValue := newPercentageValueFromFloat64(1)
+
+	for _, pa := range paths {
+		leftFractionPercentages[pa.Path] = oneHundredPercentageValue
+		leftFractionInstanceCounts[pa.Path] = 0
+	}
+
+	for _, pa := range paths {
+		endpoints, exist, err := k8sClient.GetEndpoints(namespace, pa.Backend.ServiceName)
+		if err != nil || !exist {
+			log.Warnf("fail to get endpoints %s/%s", namespace, pa.Backend.ServiceName)
+			continue
+		}
+
+		for _, subset := range endpoints.Subsets {
+			leftFractionInstanceCounts[pa.Path] += len(subset.Addresses)
```

Comment:
IMHO, this annotation will work fine with multi-port service. Since that we are doing calculations based on numbers of instances for each endpoints, it's doesn't matter which port the ingress is actually referencing.
[https://github.com/yue9944882/traefik/blob/f88a7cdd60e67070135f02655f33d16a1a220dfa/provider/kubernetes/kubernetes_test.go#L2022](https://github.com/yue9944882/traefik/blob/f88a7cdd60e67070135f02655f33d16a1a220dfa/provider/kubernetes/kubernetes_test.go#L2022) This test could be an example for multi-port service.